File: 0743-network-delay-time/0743-network-delay-time.py
class Solution:    
    def networkDelayTime(self, times, n, k):
        graph = defaultdict(list)
        for u, v, w in times:
            graph[u].append((v, w))
        
        dist = [10000] * (n+1)
        dist[k] = 0
        prQ = []
        heappush(prQ, (0, k))

        while prQ:
            total_weight, u = heappop(prQ)
            
            for v, w in graph[u]:
                if dist[u] + w < dist[v]:
                    dist[v] = dist[u]+w
                    heappush(prQ, (dist[v], v))
        
        max_d = max(dist[1:])
        return -1 if max_d==10000 else max_d

File: 0743-network-delay-time/0743-network-delay-time.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        map<int, vector<pair<int, int>>> graph;
        for (auto& v:times) {
            graph[v[0]].push_back(make_pair(v[1],v[2]));
        }
        vector<int> dist(n+1, 10000);
        priority_queue<pair<int,int>, vector<pair<int, int>>, greater<pair<int, int>>> prQ;
        prQ.push(make_pair(0, k));
        dist[k] = 0;
        int max_w = INT_MIN;

        while (!prQ.empty()) {
            int total_weight = prQ.top().first;
            int u = prQ.top().second;
            prQ.pop();

            for (const auto& edge:graph[u]) {
                int v = edge.first;
                int weight = edge.second;
                if (dist[u] + weight < dist[v]) {
                    dist[v] = dist[u]+weight;
                    prQ.push(make_pair(dist[v],v));
                }
            }
        }
        for (int i=1; i<=n; i++) {
            if ((i!=k) && dist[i] > max_w) max_w = dist[i];
        }
        return max_w==10000 ? -1 : max_w;
    }   
};